### PR TITLE
Dashboard: Improve story count for authors

### LINKED ISSUE #6613: Dashboard: author does not see drafts
## Bug Description

If I log in to an established wordPress account as a user with no stories of my own but another user has created stories, those stories aren't showing up in "my stories" (because they aren't mine) however the header counts for existing stories are showing with the number of stories that exist for that other user.  

## Expected Behaviour

If I have no stories that are getting returned to me I would expect the story counts to all be 0 regardless of what other account users have created.

## Steps to Reproduce

Log into a wordPress account that has stories created in it but as either a new users or a user with 0 stories authored by them, go to the dashboard. See header counts. 

## Additional Context

returned headers from `web-stories/v1/web-story/` are: `"headers":{"X-WP-Total":18,"X-WP-TotalPages":1,"X-WP-TotalByStatus":"{\"all\":20,\"publish\":0,\"future\":0,\"draft\":18,\"private\":2}"}}`

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

That header counts are 0 when there are no stories to show.

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Fix the counts, check if user can ability to view other posts.

## User-facing changes

| Before | After |
| ----------- | ----------- |
| Screenshot 2021-09-22 at 15 38 01 | Screenshot 2021-09-22 at 15 39 38 |
| Screenshot 2021-09-22 at 15 37 34 | Screenshot 2021-09-22 at 15 39 52 |
| Screenshot 2021-09-22 at 15 37 48 | Screenshot 2021-09-22 at 15 40 40 |

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Login as admin. 
2. Create a number of stories, with different post status, published, private, draft and scheduled. 
3. Login as author
4. Create a number of stories, with different post status, published, private, draft and scheduled. 
5. Go to dashboard, sense the number next to the filters match to the number stories displays. 
6. Ensure spinner is not shown. 
7. Login as admin. 
8. Go to dashboard, sense the number next to the filters match to the number stories displays. 
9. Ensure spinner is not shown. 

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6613

### CODE REVIEW

**Review by BrittanyIRL — APPROVED** [review]
Awesome! This resolves all of the discrepancies in story counts as well as the 'loading' status that wouldn't go away in some edge cases.

File: includes/REST_API/Stories_Controller.php [hunk 1]
```diff
@@ -426,8 +426,15 @@ public function get_items( $request ) {
...
 			$query_args['post_status'] = $status;
+			if ( in_array( $status, [ 'draft', 'future' ], true ) && ! $edit_others_posts ) {
+				$query_args['author'] = get_current_user_id();
+			}
+			if ( 'private' === $status && ! $edit_private_posts ) {
+				$query_args['author'] = get_current_user_id();
+			}
 			$posts_query->query( $query_args );
 			$statuses_count[ $key ] = absint( $posts_query->found_posts );
+			$statuses_count['all']  = $statuses_count['all'] + $statuses_count[ $key ];
```
Comment (swissspidy) [inline includes/REST_API/Stories_Controller.php]: Nit:

```suggestion
			$statuses_count['all'] += $statuses_count[ $key ];
```

File: includes/REST_API/Stories_Controller.php [hunk 2]
```diff
@@ -426,8 +426,15 @@ public function get_items( $request ) {
 		foreach ( $statuses as $key => $status ) {
 			$posts_query               = new WP_Query();
 			$query_args['post_status'] = $status;
+			if ( in_array( $status, [ 'draft', 'future' ], true ) && ! $edit_others_posts ) {
+				$query_args['author'] = get_current_user_id();
+			}
+			if ( 'private' === $status && ! $edit_private_posts ) {
+				$query_args['author'] = get_current_user_id();
+			}
```
Comment (swissspidy) [inline includes/REST_API/Stories_Controller.php:439]:
```suggestion
			if ( in_array( $status, [ 'draft', 'future', 'private' ], true ) && ! $edit_others_posts ) {
				$query_args['author'] = get_current_user_id();
			}
```

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 3]
```diff
@@ -232,14 +232,43 @@ public function test_get_items_contributor() {
...
+
+		$statuses = json_decode( $headers['X-WP-TotalByStatus'], true );
+
+		$this->assertArrayHasKey( 'all', $statuses );
+		$this->assertArrayHasKey( 'publish', $statuses );
+		$this->assertArrayHasKey( 'draft', $statuses );
+		$this->assertArrayHasKey( 'future', $statuses );
+		$this->assertArrayHasKey( 'private', $statuses );
+
+		$this->assertEquals( 10, $statuses['all'] );
```
Comment (swissspidy) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]: Let's use `assertSame` for these assertions for additional type safety, same goes for the test above.

**Comment by swissspidy** [comment]
**Before this PR:**

Screenshot 2021-09-09 at 12 31 01

**After:**

I now get an infinite loading spinner on the dashboard as an author:

Screenshot 2021-09-09 at 12 27 39

The count above also says there are 4 total stories (3 published, 1 draft), but I see only three in the list below.

There are actually 3 published stories on my site (2 by an admin, 1 by an editor).

As you can see from the screenshots, I no longer see the published "locking test" story in the dashboard.

Well, actually, I sort of see it, but it then disappears. See this video:

https://user-images.githubusercontent.com/841956/132670665-b3a537dd-09f5-45bd-bcd8-bedd877b7be5.mov

**tl;dr** something's still not quite right

File: includes/REST_API/Stories_Controller.php (hunk 2 above)
Comment (spacedmonkey) [reply includes/REST_API/Stories_Controller.php:439]: Can't do this, as `$edit_private_posts` is different from `$edit_others_posts `

**Comment by spacedmonkey** [comment]
> tl;dr something's still not quite right

@BrittanyIRL Can you take a look at this PR. 

The PHP is valid, but sure about the infinite scroll piece.

**Comment by BrittanyIRL** [comment]
> > tl;dr something's still not quite right
> 
> @BrittanyIRL Can you take a look at this PR.
> 
> The PHP is valid, but sure about the infinite scroll piece.

weird, i didn't see that when i checked - think i did the same steps. I'll check this am and get back to you.

**Comment by swissspidy** [comment]
> > tl;dr something's still not quite right
> 
> @BrittanyIRL Can you take a look at this PR.
> 
> The PHP is valid, but sure about the infinite scroll piece.

@spacedmonkey Are you able to reproduce the issue then?

**Comment by spacedmonkey** [comment]
> @spacedmonkey Are you able to reproduce the issue then?

I wasnt able to see it before, but now I can now. I also dont understand why a story appears and then disappears. 

https://user-images.githubusercontent.com/237508/132729689-c349671b-aa5a-4570-a6b5-65a580038445.mov

**Comment by BrittanyIRL** [comment]
Well, no matter what I do it's working for me. I've logged in as author, editor, and have admin editing a story at the same time. I can toggle between all stories, drafts, published, private. where necessary and it's fine. 

Screen Shot 2021-09-09 at 10 15 51 AM

It's also fine even when there's no private stories available. 
Screen Shot 2021-09-09 at 10 18 24 AM

**Comment by BrittanyIRL** [comment]
ok did a little more stress testing - upped the story count for admin > 100 and made sure author has > 50 so that it has to paginate. 

now I'm getting that issue where all the stories have disappeared for drafts and 'all stories' for an author. 

So these stories exist, i can filter down to the author drafts while logged in as an admin: 
Screen Shot 2021-09-09 at 10 29 45 AM

But then when I am logged in as the author, the drafts won't load
https://user-images.githubusercontent.com/10720454/132734262-f04e0a15-d662-40fe-808a-631dcd20e569.mp4

Screen Shot 2021-09-09 at 10 27 25 AM

This makes me thing something is going on with the pagination.

**Comment by spacedmonkey** [comment]
How about changing from 24 to 100 stories loaded? Is there a reason why we are going loading 24 at a time?

**Comment by BrittanyIRL** [comment]
> How about changing from 24 to 100 stories loaded? Is there a reason why we are going loading 24 at a time?

i don't really know why the number changed would impact how the results aren't getting loaded? can change it to 100 but it's just a bigger test case. i was using it as an example of needing to paginate. I would guess that the discrepancy is occurring somewhere in there because everything is fine if there's only 1 page to load.

**Comment by swissspidy** [comment]
> > How about changing from 24 to 100 stories loaded? Is there a reason why we are going loading 24 at a time?
> 
> i don't really know why the number changed would impact how the results aren't getting loaded? can change it to 100 but it's just a bigger test case. i was using it as an example of needing to paginate. I would guess that the discrepancy is occurring somewhere in there because everything is fine if there's only 1 page to load.

Agreed. The number of loaded stories is not the issue here. Simply changing this number will not fix this bug.

**Comment by spacedmonkey** [comment]
I think I have finally fixed it. By ensuring that X-WP-Total and X-WP-TotalPages are correct, I am not seeing the infiniate scroll issues. Can everyone retest.

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 4]
```diff
@@ -235,14 +240,47 @@ public function test_get_items_contributor() {
...
-		$this->assertEquals( 7, $statuses['publish'] );
-		$this->assertEquals( 3, $statuses['future'] );
-		$this->assertEquals( 3, $statuses['draft'] );
+		$this->assertSame( 7, $statuses['all'] );
+		$this->assertSame( 7, $statuses['publish'] );
+		$this->assertSame( 0, $statuses['future'] );
+		$this->assertSame( 0, $statuses['draft'] );
 
-		$this->assertEquals( 7, $headers['X-WP-Total'] );
+		$this->assertSame( 7, $headers['X-WP-Total'] );
```
Comment (swissspidy) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]: Looks like `7` is a string here, not an int.

**Comment by swissspidy** [comment]
> I think I have finally fixed it. By ensuring that X-WP-Total and X-WP-TotalPages are correct, I am not seeing the infiniate scroll issues. Can everyone retest.

I think we're getting there!

The spinner is now gone, but I'm still not seeing the fourth story under "All Stories":

Screenshot 2021-09-13 at 09 50 56

Interestingly, I see it under "Published":

Screenshot 2021-09-13 at 09 54 14

Here's the "Private" tab for reference:

Screenshot 2021-09-13 at 09 54 43

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php (hunk 4 above)
Comment (spacedmonkey) [reply tests/phpunit/integration/tests/REST_API/Stories_Controller.php]: Which is why it was an `assertEquals`.

**Comment by spacedmonkey** [comment]
I am not seeing this issue. 

#### As admin
Screenshot 2021-09-13 at 13 21 04

#### As author
Screenshot 2021-09-13 at 13 21 22

#### As contributor
Screenshot 2021-09-13 at 13 23 52

**Comment by swissspidy** [comment]
So strange. I edited the disappearing post in QuickEdit to change the title, and now it suddenly appears for my author user. So perhaps my data there was in a weird state.

File: includes/REST_API/Stories_Controller.php [hunk 5]
```diff
@@ -249,6 +249,7 @@ public function filter_posts_clauses( $clauses, $query ) {
 	 * Retrieves a collection of web stories.
 	 *
 	 * @SuppressWarnings(PHPMD.NPathComplexity)
+	 * @SuppressWarnings(PHPMD.CyclomaticComplexity)
 	 * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
```
Comment (swissspidy) [inline includes/REST_API/Stories_Controller.php]: A strong indicator that this method should be refactored and perhaps split up into a few smaller ones.

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php (hunk 4 above)
Comment (swissspidy) [reply tests/phpunit/integration/tests/REST_API/Stories_Controller.php]: Then it should be `$this->assertSame( '7', $headers['X-WP-Total'] );`

**Review by swissspidy — APPROVED** [review]
Tests need fixing and a refactoring might be in order, but functionality-wise it looks good.

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 6]
```diff
@@ -427,7 +465,7 @@ public function test_filter_posts_by_author_display_names() {
 		$response = rest_get_server()->dispatch( $request );
 		$results  = wp_list_pluck( $response->get_data(), 'author' );
 
-		$this->assertSame(
+		$this->assertEquals(
```
Comment (swissspidy) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]: I don't understand why this was changed again. `assertSame` should be used wherever possible for type checks

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 7]
```diff
@@ -427,7 +469,7 @@ public function test_filter_posts_by_author_display_names() {
...
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]:
```suggestion
		$this->assertSame(
```

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 8]
```diff
@@ -262,10 +303,10 @@ public function test_get_item() {
 		$data     = $response->get_data();
 		$this->assertArrayHasKey( 'preview_link', $data );
 		$view_link = get_preview_post_link( $story );
-		$this->assertSame( $view_link, $data['preview_link'] );
+		$this->assertEquals( $view_link, $data['preview_link'] );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]:
```suggestion
		$this->assertSame( $view_link, $data['preview_link'] );
```

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 9]
```diff
@@ -262,10 +303,10 @@ public function test_get_item() {
...
 		$this->assertArrayHasKey( 'edit_link', $data );
 		$edit_link = get_edit_post_link( $story, 'rest-api' );
-		$this->assertSame( $edit_link, $data['edit_link'] );
+		$this->assertEquals( $edit_link, $data['edit_link'] );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]:
```suggestion
		$this->assertSame( $edit_link, $data['edit_link'] );
```

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 10]
```diff
@@ -318,7 +359,7 @@ public function test_get_item_future() {
 		$data = $response->get_data();
 		$this->assertArrayHasKey( 'preview_link', $data );
 		$this->assertNotEmpty( $data['preview_link'] );
-		$this->assertSame( $permalink, $data['preview_link'] );
+		$this->assertEquals( $permalink, $data['preview_link'] );
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]:
```suggestion
		$this->assertSame( $permalink, $data['preview_link'] );
```

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 11]
```diff
@@ -448,7 +490,7 @@ public function test_filter_posts_by_author_display_names() {
 		$response = rest_get_server()->dispatch( $request );
 		$results  = wp_list_pluck( $response->get_data(), 'author' );
 
-		$this->assertSame(
+		$this->assertEquals(
```
Comment (spacedmonkey) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]:
```suggestion
		$this->assertSame(
```

File: includes/REST_API/Stories_Controller.php [hunk 12]
```diff
@@ -270,128 +289,137 @@ public function get_items( $request ) {
...
-		// Set before into date query. Date query must be specified as an array of an array.
-		if ( isset( $registered['before'], $request['before'] ) ) {
-			$args['date_query'][0]['before'] = $request['before'];
-		}
+	/**
+	 * Prepares the 'tax_query' for a collection of posts.
+	 *
+	 * Remove this method, once WP 5.7 is min required.
+	 *
+	 * @SuppressWarnings(PHPMD.NPathComplexity)
```
Comment (spacedmonkey) [inline includes/REST_API/Stories_Controller.php:310]: This method is copied from core. I am not going to refactor it. So this can not be avoided.

File: includes/REST_API/Stories_Controller.php [hunk 13]
```diff
@@ -270,128 +289,137 @@ public function get_items( $request ) {
...
 			}
 		}
 
+		return $args;
+	}
+
+	/**
+	 * Add response headers, with post counts.
+	 *
+	 * @SuppressWarnings(PHPMD.NPathComplexity)
```
Comment (spacedmonkey) [inline includes/REST_API/Stories_Controller.php:392]: Can't workout how to refactor this method to get this number down. Any ideas?

File: includes/REST_API/Stories_Controller.php [hunk 14]
```diff
@@ -270,128 +289,137 @@ public function get_items( $request ) {
...
+	}
 
-		// Set before into date query. Date query must be specified as an array of an array.
-		if ( isset( $registered['before'], $request['before'] ) ) {
-			$args['date_query'][0]['before'] = $request['before'];
-		}
+	/**
+	 * Prepares the 'tax_query' for a collection of posts.
+	 *
+	 * Remove this method, once WP 5.7 is min required.
```
Comment (swissspidy) [inline includes/REST_API/Stories_Controller.php]:
```suggestion
	 * @todo Remove this method once WordPress 5.7 becomes minimum required version.
```

File: includes/REST_API/Stories_Controller.php (hunk 13 above)
Comment (swissspidy) [reply includes/REST_API/Stories_Controller.php:392]: Splitting up into smaller methods is an easy way. But I'd be OK with this being done in a new ticket under the #5066 epic.

Comment (spacedmonkey) [reply includes/REST_API/Stories_Controller.php:392]: I did look into refactoring this method. It has a lot of different code paths, because of the permissions model of WordPress. I am not sure how we change this without making the method unreadable later.

File: includes/REST_API/Stories_Controller.php [hunk 15]
```diff
@@ -270,128 +289,137 @@ public function get_items( $request ) {
...
+	}
 
-		// Set before into date query. Date query must be specified as an array of an array.
-		if ( isset( $registered['before'], $request['before'] ) ) {
-			$args['date_query'][0]['before'] = $request['before'];
-		}
+	/**
+	 * Prepares the 'tax_query' for a collection of posts.
+	 *
+	 * @todo Remove this method once WordPress 5.7 becomes minimum required version.
```
Comment (BrittanyIRL) [inline includes/REST_API/Stories_Controller.php:308]: Does this need an associated ticket for clean up? We do this sometimes on our pod. Just curious.

**Review by BrittanyIRL — COMMENTED** [review]
I'm still not seeing `pending review` contributor stories show up in the dashboard. 

Screen Shot 2021-09-14 at 9 38 59 AM
Screen Shot 2021-09-14 at 9 38 47 AM

File: includes/REST_API/Stories_Controller.php (hunk 15 above)
Comment (spacedmonkey) [reply includes/REST_API/Stories_Controller.php:308]: No, this is just a note, when we bump min requirements later down the road, to remove this method.

**Comment by swissspidy** [comment]
Totally forgot about the pending post status! It's not something the dashboard currently caters for:

https://github.com/google/web-stories-wp/blob/c3dbff3bb3c3b2f7bb288c2a6c9c2c0b9406ef05/packages/dashboard/src/constants/stories.js#L104-L139

Adding support for that sounds trivial. Perhaps in a new PR?

File: includes/REST_API/Stories_Controller.php [hunk 16]
```diff
@@ -270,128 +289,137 @@ public function get_items( $request ) {
...
-			'all'     => [ 'publish' ],
 			'publish' => 'publish',
 		];
 
 		if ( $this->get_post_type_cap( $this->post_type, 'edit_posts' ) ) {
-			$statuses['all'][]  = 'draft';
-			$statuses['all'][]  = 'future';
 			$statuses['draft']  = 'draft';
 			$statuses['future'] = 'future';
 		}
```
Comment (swissspidy) [inline includes/REST_API/Stories_Controller.php:410]: Should we use `get_post_stati()` here to make this more generic?

Comment (spacedmonkey) [reply includes/REST_API/Stories_Controller.php:410]: How would that work?

Comment (swissspidy) [reply includes/REST_API/Stories_Controller.php:410]: I don't know, I haven't looked at this in detail.

This just came to mind in #9044.

Instead of hardcoding all possible post statuses here, we might be able to loop through `get_post_stati()` to make this more dynamic and DRY

**Review by BrittanyIRL — APPROVED** [review]

File: includes/REST_API/Stories_Controller.php [hunk 17]
```diff
@@ -246,10 +253,34 @@ public function filter_posts_clauses( $clauses, $query ) {
...
+		$this->args = $this->prepare_tax_query( $args, $request );
+
+		if ( 'edit' === $request['context'] ) {
+			$edit_others_posts  = $this->get_post_type_cap( $this->post_type, 'edit_others_posts' );
+			$edit_private_posts = $this->get_post_type_cap( $this->post_type, 'edit_private_posts' );
+
+			if ( array_intersect( $args['post_status'], [ 'publish', 'draft', 'future' ] ) && ! $edit_others_posts ) {
+				$args['author'] = get_current_user_id();
+			}
+			if ( in_array( 'private', $args['post_status'] ) && ! $edit_private_posts ) {
```
Comment (spacedmonkey) [inline includes/REST_API/Stories_Controller.php]:
```suggestion
			if ( in_array( 'private', $args['post_status'], true ) && ! $edit_private_posts ) {
```

File: includes/REST_API/Stories_Controller.php [hunk 18]
```diff
@@ -270,140 +289,160 @@ public function get_items( $request ) {
...
 
 		foreach ( $statuses as $key => $status ) {
 			$posts_query               = new WP_Query();
 			$query_args['post_status'] = $status;
+			if ( in_array( $status, [ 'draft', 'future' ], true ) && ! $edit_others_posts ) {
+				$query_args['author'] = get_current_user_id();
+			}
+			if ( 'private' === $status && ! $edit_private_posts ) {
+				$query_args['author'] = get_current_user_id();
+			}
```
Comment (swissspidy) [inline includes/REST_API/Stories_Controller.php:439]: Can we remove the author filtering again as per our discussion?

Users should see all stories by default, not only their own ones.

We can then create a follow-up ticket to further improve this.

I'll try to take a look at the WP core code you shared too so I get a better picture of this.

Comment (spacedmonkey) [reply includes/REST_API/Stories_Controller.php:439]: These filters are only for the counts and not displaying the story. The filter of quests was removed when I remove this commit - https://github.com/google/web-stories-wp/commit/7ff521af114ee21ea033b4fa0b595bb3bc5fcb6b

**Comment by swissspidy** [comment]
As discussed offline, merging this for now, even though there are still some edge cases to be solved.

Specifically, the issue described at https://github.com/google/web-stories-wp/issues/8532#issuecomment-907217461 still occurs.

The reason is this code in WP core that skips posts the user doesn't have permission for:

https://github.com/WordPress/wordpress-develop/blob/48c68e5dac074b4b5b059180c0715c820efd1e45/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L362-L369
